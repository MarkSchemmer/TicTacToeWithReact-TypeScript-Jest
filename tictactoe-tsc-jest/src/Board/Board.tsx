import * as React from 'react'
import Square from '../Square/Square'
import * as _ from '../Helpers/helpers'
interface IProps  {
    squares : Array<Array<_.Square>>
}

class Board extends React.Component<IProps> {
    render() {
        const { squares } = this.props
        return (
            <div>
                { squares.map((level) => level.map((sublevel) => {
                        return <Square
                                val={sublevel.val} 
                                coordinates={sublevel.coordinates}  /> 
                }))}          
            </div>
        )
    }
}

export default Board 