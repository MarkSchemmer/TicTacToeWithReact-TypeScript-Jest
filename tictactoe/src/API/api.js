

export const OPTIONS = {
    XWIN : 1,
    XLOSE : 2,
    TIE : 3, 
    MOREMOVES : 4
}


let threeInRow = (chr, [a,b,c], nb) =>  nb[a] === nb[b] && nb[b] === nb[c] && nb[a] === chr

export function IsWinner (board){

    let nb = board
    .reduce((acc,cur) => acc.concat(cur), [])
    .map( x => x.val )

    console.log(nb)

    let waysToWin = [
        [0,1,2],
        [0,4,8],
        [0,3,6],
        [1,4,7],
        [2,4,6],
        [2,5,8],
        [3,4,5],
        [6,7,8]
    ]

    for(let i = 0; i < waysToWin.length; i++) {

        console.log(threeInRow('X', waysToWin[i], nb))

        if(threeInRow('X', waysToWin[i], nb))
            return OPTIONS.XWIN
        else if (threeInRow('O', waysToWin[i], nb))
            return OPTIONS.XLOSE
    }

    return nb.every(x => x !== null ) ? OPTIONS.TIE : OPTIONS.MOREMOVES
}
