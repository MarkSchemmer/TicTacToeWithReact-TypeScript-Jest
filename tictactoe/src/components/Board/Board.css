
.row { 
    width:400px;
    height: 100px;
    /* border: 1px solid black; */
    margin: auto;
}


.column { 
    width: 130px;
    height: 100px;
    border : 1px solid black;
    display: inline-block;
}