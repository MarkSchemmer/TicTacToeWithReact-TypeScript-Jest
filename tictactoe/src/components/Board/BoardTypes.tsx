import React, { Component } from 'react'


interface ISquareAttributes {
    id : number, 
    coor : Array<number> 
    val : string | null 
}

interface IBoardWithMove {
    history : Array<Array<ISquareAttributes>>,
    Move : Array<Array<number>> | null,
    WhomMoved : string | null 
}


export class SquareAttributes implements ISquareAttributes {
    public id : number; 
    public coor : Array<number>;
    public val : string | null;

    constructor(id : number, coor : Array<number>, val : string | null = null){
        this.id = id
        this.coor = coor 
        this.val = val 
    }
}

export class BoardWithMoves implements IBoardWithMove {
    
    public history : Array<Array<SquareAttributes>>

    public Move : Array<Array<number>>

    public WhomMoved : string | null 
    
    constructor(history:Array<Array<SquareAttributes>>,
         Move : Array<Array<number>>, 
         WhomMoved : string | null = null ){
        this.history = history
        this.Move = Move 
        this.WhomMoved = WhomMoved
    }
}