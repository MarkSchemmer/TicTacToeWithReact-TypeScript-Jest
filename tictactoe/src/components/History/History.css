
.history {
    position: absolute;
}

.history button {
    /* display: block; */
    position: relative;
    left : 200px;
    display: flex;
    flex-direction: column;
}