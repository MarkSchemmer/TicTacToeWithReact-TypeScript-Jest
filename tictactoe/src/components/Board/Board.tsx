import React, { Component } from 'react'
import History from '../History/History'
import Square from '../Square/Square'
import './Board.css'

import * as api from '../../API/api'

import  * as BoardAttributes from './BoardTypes'

interface IProps {

}

interface IState {
    name : string, 
    board : Array<BoardAttributes.BoardWithMoves>,
    turn : number,
    IsWinner : number | null,
}

class Board extends Component<IProps, IState> {


    public state : IState = {
        name : 'Mark',
        board : [api.genBoard()],
        turn : 0,
        IsWinner : null, 
    }


    public onClick = (coor:Array<number>) => {
        const [x,y] = coor

        if(this.state.board[this.state.turn].history[x][y].val !== null) return  

        if(this.state.IsWinner !== null) return 

        // copy of history
        let copyOfHistory = this.state.board[this.state.turn].history

        // copy of Move 
        let copyOfMove = this.state.board[this.state.turn].Move 
        copyOfMove = [...copyOfMove, [x,y]]

        let whosMoving
    

        if(this.state.turn%2==0)
            whosMoving = 'X'
        else 
            whosMoving = 'O'

        copyOfHistory[x][y].val = whosMoving

        let newCopy = new BoardAttributes.BoardWithMoves(copyOfHistory, copyOfMove, whosMoving)

        this.setState((prevState) => {
            return {
                turn : prevState.turn+1,
                board : [...this.state.board, newCopy]
            }
        }, () => this.checkForWinner())
    }
 

    public checkForWinner = () => {
        console.log('the board: ',this.state.board)
        const hasWinner = api.IsWinner(this.state.board[this.state.turn].history)
        let whichOption = null 

        if(hasWinner===api.OPTIONS.XWIN)
            whichOption = api.OPTIONS.XWIN
        else if (hasWinner === api.OPTIONS.XLOSE) 
              whichOption = api.OPTIONS.XLOSE
        else if (hasWinner === api.OPTIONS.TIE)
                whichOption = api.OPTIONS.TIE

     if (whichOption)
        this.setState( { IsWinner : whichOption }, () => {
            this.setWinner()
        }) 
    }

    public setWinner = () => {
        if(this.state.IsWinner !== null){

            if(this.state.IsWinner===api.OPTIONS.TIE){
                return (
                    <React.Fragment>
                        TIE GAME!
                    </React.Fragment>
                )
            } else {
                if(this.state.IsWinner===api.OPTIONS.XWIN){
                    return (
                        <React.Fragment>
                            Player X You Win!
                        </React.Fragment>
                    )
                } else {
                   return ( <React.Fragment>
                        Player O You Win!
                   </React.Fragment> )
                }
            }

        } else { 
            return (
                <React.Fragment>
                    Player { this.state.turn % 2 === 0 ? <strong>{'X'}</strong> : <strong>{'O'}</strong> } your move!
                </React.Fragment>
            )
        }
    }

    public render () {
        return (
            <React.Fragment>
                <div className="main">
                <History history={this.state.board} /> 
                  <div className="greetings">Greetings { this.state.name }</div>
                  <div className="whoms-turn">  { this.setWinner() } </div>
                    <div className="board">
                            { this.state.board[this.state.turn].history.map((x:any) => 
                               <div className="row" key={ Date.now() * (Math.random() * 100) }>
                                    { x.map((z:any) => <Square key={z.id} {...z} 
                                        onClick={this.onClick.bind(this)} /> )}
                               </div> )}
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

export default Board 

