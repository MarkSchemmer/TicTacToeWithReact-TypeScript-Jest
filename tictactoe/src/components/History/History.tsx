import React, { Component } from 'react'
import * as  BoardAttributes from '../Board/BoardTypes'
import './History.css'

interface IProps {
    history : Array<BoardAttributes.BoardWithMoves>
}

class History extends Component<IProps> {

    constructor(props : IProps) {
        super(props)
    }


        public render () {
            return (
                    <div className="history">
                            { this.props.history.map((obj, index) => 
                                
                                    {if(index === 0) {
                                            return (
                                                <React.Fragment key={index}>
                                                    <button>Start of Match</button>
                                                </React.Fragment>
                                            )
                                    } else {
                                        let [x,y] = obj.Move
                                            return (
                                                <React.Fragment key={index*Date.now()}>
                                                    <button>{` Player: ${obj.WhomMoved} Move :${index} Coordinate : (${x},${y}) `}</button>   
                                                </React.Fragment>
                                            )
                                    }}
                                )}
                    </div>
            )
    }
}

export default History