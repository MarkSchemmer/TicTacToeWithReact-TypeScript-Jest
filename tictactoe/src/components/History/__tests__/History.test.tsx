import * as React from 'react'
import * as enzyme from 'enzyme'
import  Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import Board from '../../Board/Board'
import History from '../History'

import * as types from '../../Board/BoardTypes'

import * as api from '../../../API/api.js'


enzyme.configure({ adapter: new Adapter() });

describe(' dummy test for History', () => {
    it('dummy it', () => {
        expect(true).toEqual(true)
    })
})


describe(' simulating clicks on board and on history buttons ', () => {
    
    it(' click board 3 time should have 4 buttons for history ', () => {
        const wrapper = enzyme.mount( <Board /> )
        let findSquares = wrapper.find('.column')
        let historyButtons = wrapper.find('.history button')
        let seriesOfClicksOnSquares = [0,1,5,2]

        seriesOfClicksOnSquares.forEach(x => {
            findSquares.at(x).simulate('click')
            wrapper.update()
        })

        expect(wrapper.find('.hist-btn')).toHaveLength(4)
    })

    /*  statement P:
        click board 3 times and then click 3rd history button button 
        should be highlight and board should show only pos 0 should
        be X and pos 1 should be O
    */

    it(' Please refer to comment statement P ', () => {
        const wrapper = enzyme.mount(<Board/> )
        let findSquares = wrapper.find('.column')
        let historyButtons = wrapper.find('.hist-btn')
        let seriesOfClicksOnSquares = [0, 1, 5, 2]

        // should be empty Baord 
       //  expect(findSquares.everyWhere(x => x.text() )).toEqual('')

        seriesOfClicksOnSquares.forEach(x => {
            findSquares.at(x).simulate('click')
            wrapper.update()
        })

        expect(wrapper.find('.hist-btn')).toHaveLength(4)

        expect(wrapper.find('.hist-btn').at(2).html()).toEqual("<button class=\"hist-btn\">  Player: X Move :3 Coordinate : (1, 2)  </button>")

        wrapper.find('.hist-btn').at(2).simulate('click')
        wrapper.update()

        expect(wrapper.find('.hist-btn').at(2).html()).toEqual("<button class=\"hist-btn\"><strong> Player: X Move :3 Coordinate : (1, 2) </strong></button>")

        let findUpdatedSquares = wrapper.find('.column')

        expect(findUpdatedSquares.at(0).text()).toEqual('X')
        expect(findUpdatedSquares.at(1).text()).toEqual('O')

        expect(findUpdatedSquares.at(2).text()).toEqual('')
        expect(findUpdatedSquares.at(3).text()).toEqual('')
        expect(findUpdatedSquares.at(4).text()).toEqual('')
        expect(findUpdatedSquares.at(5).text()).toEqual('X')
        expect(findUpdatedSquares.at(6).text()).toEqual('')
        expect(findUpdatedSquares.at(7).text()).toEqual('')
        expect(findUpdatedSquares.at(8).text()).toEqual('')


        expect(findUpdatedSquares).toHaveLength(9)
        
    })
})


describe(' first two buttons are always Toggle Moves and Start Match...', () => {

    let props, goBackInTime, board 

    beforeEach(() => {
        props = {
            history : [api.genBoard()]
        }

        goBackInTime = jest.fn()

        board = enzyme.mount( <Board /> )
    })

    it('Toggle is first', () => {
        const wrapper = enzyme.shallow( <History {...props} goBackInTime={goBackInTime} /> )

        expect(wrapper.find('button').at(0).text()).toEqual('Toggle Moves')
    })

    it(' Start Match is seconded ', () => {
        const wrapper = enzyme.shallow( <History {...props} goBackInTime={goBackInTime} /> )
        
        expect(wrapper.find('button').at(1).text()).toEqual('Start Match')
    })

    it(' simulate click on board and should have 3 buttons in history ', () => {
        board.find('.column').at(0).simulate('click')
        board.update()

        expect(board.find('.history button')).toHaveLength(3)
    })
})


describe(' Snapshot for <History /> ', () => {

    let props, goBackInTime

    beforeEach(() => {
        goBackInTime = jest.fn()
        props = {
            history : [api.genBoard()], 
        }
    })

    it(' Snapshot of <History /> ', () => {
        const tree = renderer.create( <History {...props} goBackInTime={goBackInTime} /> ).toJSON()
        expect(tree).toMatchSnapshot() 
    })
})