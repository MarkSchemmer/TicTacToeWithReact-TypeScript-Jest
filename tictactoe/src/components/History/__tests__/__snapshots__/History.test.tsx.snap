// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Snapshot for <History />   Snapshot of <History />  1`] = `
<div
  className="history"
>
  <button
    className="hist-btn"
    onClick={[Function]}
  >
     
    Start of Match
     
  </button>
</div>
`;
